javatests/arcs/core/data/proto/ParticleProtoDecoderTest.kt

```diff
@@ -0,0 +1,126 @@
+            .newBuilder()
+            .setName("unknown")
+            .setHandle("thing")
+            .build()
+        val exception = assertThrows(IllegalArgumentException::class) {
+            proto.decode(readerSpec, context)
+        }
+        assertThat(exception)
+            .hasMessageThat()
+            .contains("HandleConnectionSpec 'unknown' not found in ParticleSpec 'Reader'")
+    }
+
+    @Test
+    fun decodeHandleConnectionDetectsMissingHandleProto() {
+        val proto = HandleConnectionProto
+            .newBuilder()
+            .setName("data")
+            .setHandle("unknown")
+            .build()
+        val exception = assertThrows(IllegalArgumentException::class) {
+            proto.decode(writerSpec, context)
+        }
+        assertThat(exception)
+            .hasMessageThat()
+            .contains("HandleProto for 'unknown' not found when decoding ParticleProto 'Writer'")
+    }
+
+    @Test
+    fun decodesParticleProto() {
+        val readConnectionProto = HandleConnectionProto
+            .newBuilder()
+            .setName("data")
+            .setHandle("thing")
+            .build()
+        val particleProto = ParticleProto
+            .newBuilder()
+            .setSpecName("Reader")
+            .addConnections(readConnectionProto)
+            .build()
+        val particle = particleProto.decode(context)
+        val readConnection = readConnectionProto.decode(readerSpec, context)
+        assertThat(particle.particleName).isEqualTo("Reader")
+        assertThat(particle.location).isEqualTo(readerSpec.location)
+        assertThat(particle.handles).isEqualTo(mapOf("thing" to readConnection))
+    }
+
+    @Test
+    fun decodeParticleDetectsMissingSpecs() {
+        val particleProto = ParticleProto
+            .newBuilder()
+            .setSpecName("NonExistent")
+            .build()
+        val exception = assertThrows(IllegalArgumentException::class) {
+            particleProto.decode(context)
+        }
+        assertThat(exception).hasMessageThat().contains("ParticleSpec 'NonExistent' not found")
+    }
+
+    // The test environment.
+    var ramdiskStorageKey = "ramdisk://something"
+    val thingHandleProto = HandleProto
+        .newBuilder()
+        .setName("thing")
+        .setFate(HandleProto.Fate.CREATE)
+        .setStorageKey(ramdiskStorageKey)
+        .build()
+    val readConnectionSpec = HandleConnectionSpec("data", Direction.READS, TypeVariable("data"))
+    val readerSpec = ParticleSpec("Reader", mapOf("data" to readConnectionSpec), "ReaderLocation")
+    val writeConnectionSpec = HandleConnectionSpec("data", Direction.WRITES, TypeVariable("data"))
+    val writerSpec = ParticleSpec("Writer", mapOf("data" to writeConnectionSpec), "WriterLocation")
+    val context = DecodingContext(
+        particleSpecs = mapOf("Reader" to readerSpec, "Writer" to writerSpec),
+        handleProtos = mapOf("thing" to thingHandleProto)
+    )
```

Comment:
These can go at the top of the file (they're not in a companion object, so they're just instance fields)